File: Week_01/java/src/_83_RemoveDuplicatesFromSortedList.java
/**
 * 第一遍：2020/07/19周日 ✅
 * 第二遍：2020/07/19周二 ✅
 * 第三遍：2020/07/08周四
 * 第四遍：2020/07/13周一
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 */
class _83_RemoveDuplicatesFromSortedList {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            while (curr.next != null && curr.next.val == curr.val) {
                curr.next = curr.next.next;
            }
            curr = curr.next;
        }
        return head;
    }

    class SolutionRecursion {
        public ListNode deleteDuplicates(ListNode head) {
            if (head == null || head.next == null) return head;
            head.next = deleteDuplicates(head.next);
            return head.val == head.next.val ? head.next : head;
        }
    }

    /**
     * Definition for singly-linked list.
     */
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}

File: Week_01/kotlin/src/_83_RemoveDuplicatesFromSortedList.kt
fun deleteDuplicates(head: ListNode?): ListNode? {
    if (head == null || head.next == null) return head
    head.next = deleteDuplicates(head.next)
    return if (head.next!!.`val` == head.`val`) head.next else head
}

fun deleteDuplicates2(head: ListNode?): ListNode? {
    var curr = head
    while (curr != null) {
        while (curr.next?.`val` == curr.`val`) {
            curr.next = curr.next?.next
        }
        curr = curr.next
    }
    return head
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}